src/groth16/circuit/builder/mod.rs

```diff
@@ -287,27 +645,333 @@ where
+        mut gate: F,
+    ) -> Word8
+    where
+        F: FnMut(&mut Self, WireId, WireId) -> WireId,
+    {
+        let mut wrd8: Word8 = *inputs
+            .next()
+            .expect("u8_fan_in: input iterator must have at least one element");
+        inputs.fold(wrd8, |acc, next| {
+            acc.iter()
+                .zip(next.iter())
+                .enumerate()
+                .for_each(|(i, (&l, &r))| wrd8[i] = gate(self, l, r));
+            wrd8
+        });
+        wrd8
+    }
+
+    pub fn u64_bitwise_op<F>(&mut self, left: &Word64, right: &Word64, mut gate: F) -> Word64
+    where
+        F: FnMut(&mut Self, WireId, WireId) -> WireId,
+    {
+        let mut wrd64 = Word64::default();
+        left.iter()
+            .zip(right.iter())
+            .flat_map(|(l, r)| l.iter().zip(r.iter()))
+            .zip(iproduct!(0..8, 0..8))
+            .for_each(|((&l, &r), (i, j))| wrd64[i][j] = gate(self, l, r));
+        wrd64
+    }
+
+    pub fn u8_bitwise_op<F>(&mut self, left: &Word8, right: &Word8, mut gate: F) -> Word8
+    where
+        F: FnMut(&mut Self, WireId, WireId) -> WireId,
+    {
+        let mut wrd8 = Word8::default();
+        left.iter()
+            .zip(right.iter())
+            .enumerate()
+            .for_each(|(i, (&l, &r))| wrd8[i] = gate(self, l, r));
+        wrd8
+    }
+
+    pub fn u64_unary_op<F>(&mut self, input: &Word64, mut gate: F) -> Word64
+    where
+        F: FnMut(&mut Self, WireId) -> WireId,
+    {
+        let mut wrd64 = Word64::default();
+        input
+            .iter()
+            .flat_map(|x| x.iter())
+            .zip(iproduct!(0..8, 0..8))
+            .for_each(|(&x, (i, j))| wrd64[i][j] = gate(self, x));
+        wrd64
+    }
+
+    pub fn u8_unary_op<F>(&mut self, input: &Word8, mut gate: F) -> Word8
+    where
+        F: FnMut(&mut Self, WireId) -> WireId,
+    {
+        let mut wrd8 = Word8::default();
+        input
+            .iter()
+            .enumerate()
+            .for_each(|(i, &x)| wrd8[i] = gate(self, x));
+        wrd8
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    ///////////////////////////// Comparison Functions /////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////
+
+    // fn new_4_bit_compare(&mut self, lhs: [WireId; 4], rhs: [WireId; 4], cmp: [WireId; 3]) -> [WireId; 3] {
 
-        while literal != 0 {
-            let wire = match literal % 2 {
-                0 => self.zero_wire(),
-                1 => self.unity_wire(),
-                _ => unreachable!(),
+    // }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////// Keccak Functions ////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////
+
+    fn keccakf_1600(&mut self, a: &mut [Word64; 25]) {
```

Comment:
If I understand what you've done correctly, you've closely followed the implementation of tiny keccak. However, in the future I would like to bring the interfaces back to what we planned initially, where each circuit builder method takes in the input wires as an argument, and returns the output wires of the circuit. For instance, since each step of the keccak hash changes the state, we would have something that looked like
```
let state = circuit.new_keccak_matrix();
...
let state = circuit.keccak_theta(state);
...
let state = circuit.keccak_rho_and_pi(state);
...
```
where we are shadowing the `state` variable to avoid lots of temporary variable names.

However, internally we can in some cases use the mutating of a mutable reference when we are not actually chaning the structure of the circuit, such as when we are rotating the wires. But I would still want the `rho_and_pi` step to adhere to the standard interface, even though it only rotates the state wires, because it's more of a higher level function in the keccak hash.

Reply:
I would normally agree, but after writing countless implementations of the parts of keccakf and trying desperately to get them to behave the same as tiny_keccak I doubt we will ever be able to accomplish this. In all honesty, I think the simplifications tiny_keccak makes (not sure which) are Not equivalent to the specification of keccak. I could be wrong, but I never was able to write rust code (not even without circuits) that was based on the pseudo code which behaved the same as tiny_keccak.